// @dainprotocol/service-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DAINServicePayments = exports.DAINPlugins = exports.CoreUtils = exports.createToolbox = exports.createService = exports.createTool = exports.DAINServiceClient = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./service"), exports);
exports.DAINServiceClient = tslib_1.__importStar(require("./client"));
// Explicitly re-export the utility functions
var service_1 = require("./service");
Object.defineProperty(exports, "createTool", { enumerable: true, get: function () { return service_1.createTool; } });
Object.defineProperty(exports, "createService", { enumerable: true, get: function () { return service_1.createService; } });
Object.defineProperty(exports, "createToolbox", { enumerable: true, get: function () { return service_1.createToolbox; } });
Object.defineProperty(exports, "CoreUtils", { enumerable: true, get: function () { return service_1.CoreUtils; } });
// Export plugins for easy access
exports.DAINPlugins = tslib_1.__importStar(require("./plugins"));
// If you want to keep the previous naming convention, you can do:
// export * as DAINService from './service';
// This will create a namespace DAINService containing all exports from './service'
exports.DAINServicePayments = tslib_1.__importStar(require("./lib/payments"));
